### ISSUE TITLE
Clang++ 10 build fails

### ISSUE THREAD

**Issue by eugeneswalker** [zeromq/libzmq#3917]
# Issue description

Building `libzmq 4.3.2` fails using `clang++ 10.0` on `Ubuntu 18.04` for `x86_64`

```
$> ./configure --with-libsodium CFLAGS=-Wno-gnu CXXFLAGS=-Wno-gnu
$> make
...
In file included from src/mtrie.cpp:32:
./src/generic_mtrie_impl.hpp:52:46: error: ISO C++ requires the name after '::~' to be found in the same scope as the name before '::~' [-Werror,-Wdtor-name]
template <typename T> zmq::generic_mtrie_t<T>::~generic_mtrie_t ()
                      ~~~~~~~~~~~~~~~~~~~~~~~^~
                                             ::generic_mtrie_t
1 error generated.
Makefile:4468: recipe for target 'src/src_libzmq_la-mtrie.lo' faile
```

**Comment by bluca** [comment]
Does it work if you don't use -Wno-gnu?

**Comment by eugeneswalker** [comment]
@bluca same error if I do not use `-Wno-gnu`

**Comment by eugeneswalker** [comment]
I was able to get the build to work using configure arg `--disable-Werror` , as the default configure options enable `-Werror` which was turning the warning `-Wdtor-name` into an error. Should we close this issue? Thanks!

**Comment by bluca** [comment]
no, the warning should be fixed

**Comment by bluca** [comment]
Cannot reproduce this with clang10 from sid.

**Comment by bluca** [comment]
It happens on oss-fuzz builds too - @sigiesec @somdoron any idea? This is pure templater-y witchcraft

**Comment by eugeneswalker** [comment]
> Cannot reproduce this with clang10 from sid.

What is sid?

**Comment by gummif** [comment]
Looks like the implementation should be inside zmq namespace scope, see https://godbolt.org/z/Pqk6ba

**Comment by bluca** [comment]
Looks like you are right, thanks for the tip

**Comment by gruenich** [comment]
After reading https://bugs.llvm.org/show_bug.cgi?id=46979, I think the proper fix is reverting #3922 and altering line 52 to
```
template <typename T> zmq::generic_mtrie_t<T>::~generic_mtrie_t<T> ()
```

**Comment by yves-surrel** [comment]
I have the same problem (building through buildozer with MacOS 11.7). Setting `--disable-Werror` allows to compile, but in use the library send cryptic errors (using adb logcat), the same with the change suggested by gruenich.
It would be nice to have this fixed.
